sql/atxs/atxs.go

```diff
@@ -0,0 +1,208 @@
+}
+
+// GetTimestampByID gets an ATX timestamp by a given ATX ID.
+func GetTimestampByID(db sql.Executor, id types.ATXID) (timestamp time.Time, err error) {
+	enc := func(stmt *sql.Statement) {
+		stmt.BindBytes(1, id.Bytes())
+	}
+	dec := func(stmt *sql.Statement) bool {
+		timestamp = time.Unix(0, stmt.ColumnInt64(0))
+		return true
+	}
+
+	if rows, err := db.Exec("select timestamp from atxs where id = ?1;", enc, dec); err != nil {
+		return time.Time{}, fmt.Errorf("exec id %v: %w", id, err)
+	} else if rows == 0 {
+		return time.Time{}, fmt.Errorf("exec id %s: %w", id, sql.ErrNotFound)
+	}
+
+	return timestamp, err
+}
+
+// GetLastIDByNodeID gets the last ATX ID for a given node ID.
+func GetLastIDByNodeID(db sql.Executor, nodeID types.NodeID) (id types.ATXID, err error) {
+	enc := func(stmt *sql.Statement) {
+		stmt.BindBytes(1, nodeID.ToBytes())
+	}
+	dec := func(stmt *sql.Statement) bool {
+		stmt.ColumnBytes(0, id[:])
+		return true
+	}
+
+	if rows, err := db.Exec(`
+		select id from atxs 
+		where smesher = ?1
+		order by epoch desc 
+		limit 1;`, enc, dec); err != nil {
+		return types.ATXID{}, fmt.Errorf("exec id %v: %w", id, err)
+	} else if rows == 0 {
+		return types.ATXID{}, fmt.Errorf("exec id %s: %w", id, sql.ErrNotFound)
+	}
+
+	return id, err
+}
+
+// GetIDByEpochAndNodeID gets an ATX ID for a given epoch and node ID.
+func GetIDByEpochAndNodeID(db sql.Executor, epoch types.EpochID, nodeID types.NodeID) (id types.ATXID, err error) {
+	enc := func(stmt *sql.Statement) {
+		stmt.BindInt64(1, int64(epoch))
+		stmt.BindBytes(2, nodeID.ToBytes())
+	}
+	dec := func(stmt *sql.Statement) bool {
+		stmt.ColumnBytes(0, id[:])
+		return true
+	}
+
+	if rows, err := db.Exec(`
+		select id from atxs 
+		where epoch = ?1 and smesher = ?2 
+		limit 1;`, enc, dec); err != nil {
+		return types.ATXID{}, fmt.Errorf("exec id %v: %w", id, err)
+	} else if rows == 0 {
+		return types.ATXID{}, fmt.Errorf("exec id %s: %w", id, sql.ErrNotFound)
+	}
+
+	return id, err
+}
+
+// GetIDsByEpoch gets ATX IDs for a given epoch.
+func GetIDsByEpoch(db sql.Executor, epoch types.EpochID) (ids []types.ATXID, err error) {
+	enc := func(stmt *sql.Statement) {
+		stmt.BindInt64(1, int64(epoch))
+	}
+	dec := func(stmt *sql.Statement) bool {
+		var id types.ATXID
+		stmt.ColumnBytes(0, id[:])
+		ids = append(ids, id)
+		return true
+	}
+
+	if rows, err := db.Exec("select id from atxs where epoch = ?1;", enc, dec); err != nil {
+		return nil, fmt.Errorf("exec epoch %v: %w", epoch, err)
+	} else if rows == 0 {
+		return []types.ATXID{}, nil
+	}
+
+	return ids, nil
+}
+
+// GetTop gets a top ATX (positioning ATX candidate) for a given ATX ID.
+// The top ATX is the one with the highest layer ID.
+func GetTop(db sql.Executor) (id types.ATXID, err error) {
+	dec := func(stmt *sql.Statement) bool {
+		stmt.ColumnBytes(0, id[:])
+		return true
+	}
+
+	if rows, err := db.Exec("select id from atxs order by layer desc, timestamp asc limit 1;", nil, dec); err != nil {
```

Comment:
this is too inneficient. it will scan all atxs and then sort them. we need a better solution.
it is also bad to add index for the layer... as every index will carry an additional cost and the query is very simple.

Reply:
what about a separate table for the top ATX?

Reply:
the best i can think of is to add `label` field to the atx table. one of labels will be `top` (just some integer). after Add`ing atx to the table in another action (UpdateTop) we will check if new atx is higher than the current `top` and if it is we remove the label from the last one and add it to the new one. `label` is indexed and is unique. Add and UpdateTop are executed as a part of the same transaction

Reply:
> what about a separate table for the top ATX?

this works as well, but i think label may be cleaner

Reply:
I tried to implement the approach you suggested but I didn't like the complexity of queries and I think it would be less performant than a separate table. So I decided to choose the approach with a separate table. I hope you don't mind